core/src/test/java/com/linecorp/armeria/client/endpoint/healthcheck/PartialHealthCheckStrategyTest.java

```diff
@@ -0,0 +1,279 @@
+    private List<Endpoint> candidatesForMaxRatio;
+    private List<Endpoint> candidatesForMaxCount;
+
+    private static List<Endpoint> createCandidates(int size) {
+        final Random random = new Random();
+
+        return IntStream.range(0, size)
+                        .mapToObj(i -> Endpoint.of("dummy" + random.nextInt()))
+                        .collect(Collectors.toList());
+    }
+
+    private static void assertCandidates(List<Endpoint> act, List<Endpoint> exp) {
+        assertThat(act).hasSize(exp.size());
+
+        for (Endpoint expCandidate : exp) {
+            assertThat(act.contains(expCandidate)).isTrue();
+        }
+    }
+
+    @BeforeEach
+    void beforeEach() {
+        maxRatioStrategy = new PartialHealthCheckStrategyBuilder().maxEndpointRatio(MAX_RATIO).build();
+        maxCountStrategy = new PartialHealthCheckStrategyBuilder().maxEndpointCount(MAX_COUNT).build();
+
+        candidatesForMaxRatio = createCandidates(10);
+        candidatesForMaxCount = createCandidates(6);
+
+        maxRatioStrategy.updateCandidates(candidatesForMaxRatio);
+        maxCountStrategy.updateCandidates(candidatesForMaxCount);
+    }
+
+    @Test
+    void getCandidatesWhenBeforeFirstUpdateCandidates() {
+        maxRatioStrategy = new PartialHealthCheckStrategyBuilder().maxEndpointRatio(MAX_RATIO)
+                                                                  .build();
+
+        assertThat(maxRatioStrategy.getCandidates()).isEmpty();
+    }
+
+    @Test
+    void getCandidatesWhenUpdateCandidatesByEmpty() {
+        maxRatioStrategy.updateCandidates(new ArrayList<>());
+        assertThat(maxRatioStrategy.getCandidates()).isEmpty();
+    }
+
+    @Test
+    void getCandidates() {
+        maxRatioStrategy.updateCandidates(candidatesForMaxRatio);
+
+        final List<Endpoint> actCandidates = maxRatioStrategy.getCandidates();
+        assertThat(actCandidates).hasSize(9);
+
+        actCandidates.forEach(
+                actCandidate -> assertThat(candidatesForMaxRatio.contains(actCandidate)).isTrue());
+    }
+
+    @Test
+    void updateHealthWhenEndpointIsHealthy() {
+        final Endpoint endpoint = candidatesForMaxRatio.get(1);
+
+        final boolean actRes = maxRatioStrategy.updateHealth(endpoint, HEALTHY);
+
+        assertThat(actRes).isFalse();
+    }
+
+    @Test
+    void updateHealthWhenEndpointIsUnhealthyOnMaxRatioMode() {
+        final Endpoint unhealthyEndpoint = maxRatioStrategy.getCandidates().get(0);
+
+        final boolean actRes = maxRatioStrategy.updateHealth(unhealthyEndpoint, UNHEALTHY);
+        assertThat(actRes).isTrue();
+
+        final List<Endpoint> actCandidates = maxRatioStrategy.getCandidates();
+        assertThat(actCandidates).hasSize(9);
+        assertThat(actCandidates.contains(unhealthyEndpoint)).isFalse();
+        actCandidates.forEach(
+                actCandidate -> assertThat(candidatesForMaxRatio.contains(actCandidate)).isTrue());
```

Comment:
Could you also check if the `selectedCandidates` contains distinct endpoints? i.e. no duplicate

Reply:
add test code